### ISSUE TITLE
Support for non-valid UTF-8 strings

### ISSUE THREAD

**Issue by ChuckDaniels87** [ChuckDaniels87/rnr#6]
Extend code to support non-valid UTF-8 strings in filenames, paths and arguments:
* Use OsStr and OsString.
* Follow [OsStr pattern API extension](https://github.com/rust-lang/rust/issues/49802) in Rust repository.
* Check issues with current crates: `clap`, `regex`, `walkdir` and `ansi_term`

**Comment by ChuckDaniels87** [comment]
Right now it is not possible to convert `OsStr(ing)` to `&[u8]` on Windows to be used in `regex::bytes::Regex::replace` without losing information. For example, `ripgrep` [uses a `to_string_lossy` conversion](https://github.com/BurntSushi/ripgrep/blob/c4a5bc06c5e445015a422398cfb69a34815390d4/globset/src/pathutil.rs#L93-L104) to obtain a `&[u8]` in Windows.

**Comment by BurntSushi** [comment]
Yeah, this is something I've always wondered about. So far, I haven't had anyone complain about cases where information is lost, i.e., when there's an invalid UTF-16 file path on Windows. One presumes that this might be so infrequent that it may not be a blocking problem in practice.

Getting a real fix for this is tricky. One possibility is to use the underlying representation of an `OsStr` (which is WTF-8), but this is not part of the public API. Another possibility is to re-create WTF-8 decoding outside of `std` using the Windows version of the [`OsStrExt` trait](https://doc.rust-lang.org/std/os/windows/ffi/trait.OsStrExt.html). But this incurs a second WTF-8 decoding step, however, it's no worse than the lossy UTF-8 decoding that I'm already doing.
